File: Java/src/graph/ConnectedComponent.java
package graph;

import edu.princeton.cs.algs4.*;
import edu.princeton.cs.algs4.Edge;
import edu.princeton.cs.algs4.EdgeWeightedGraph;
import edu.princeton.cs.algs4.Graph;

/**
 * 使用深度优先搜索查找图中的连通分量
 *
 * @author igaozp
 * @since 2017-07-13
 * @version 1.0
 */
public class ConnectedComponent {
    /**
     * 存储图中顶点的访问状态
     */
    private boolean[] marked;
    /**
     * 连通分量的标识符
     */
    private int[] id;
    /**
     * 存放给定分量的顶点数量
     */
    private int[] size;
    /**
     * 顶点的访问次数
     */
    private int count;

    /**
     * 构造方法
     *
     * @param G 初始化的图
     */
    public ConnectedComponent(Graph G) {
        marked = new boolean[G.V()];
        id = new int[G.V()];
        size = new int[G.V()];

        for (int s = 0; s < G.V(); s++) {
            if (!marked[s]) {
                dfs(G, s);
                count++;
            }
        }
    }

    /**
     * 构造方法
     *
     * @param G 带权重的图
     */
    public ConnectedComponent(EdgeWeightedGraph G) {
        marked = new boolean[G.V()];
        id = new int[G.V()];
        size = new int[G.V()];

        for (int v = 0; v < G.V(); v++) {
            if (!marked[v]) {
                dfs(G, v);
                count++;
            }
        }
    }

    /**
     * 深度优先搜索
     *
     * @param G 搜索的图
     * @param v 起始顶点
     */
    private void dfs(Graph G, int v) {
        marked[v] = true;
        id[v] = count;

        // 访问与 v 连通的顶点
        for (int w : G.adj(v)) {
            if (!marked[w]) dfs(G, w);
        }
    }

    /**
     * 深度优先搜索
     *
     * @param G 搜索的图
     * @param v 起始顶点
     */
    private void dfs(EdgeWeightedGraph G, int v) {
        marked[v] = true;
        id[v] = count;
        size[count]++;

        // 访问与 v 连通的顶点
        for (Edge e : G.adj(v)) {
            int w = e.other(v);
            if (!marked[w]) dfs(G, w);
        }
    }

    /**
     * 检查两个连通分量
     *
     * @param v 连通分量
     * @param w 连通分量
     * @return {@code true} 两个分量连通
     *         {@code false} 两个分量不连通
     */
    public boolean connected(int v, int w) {
        validateVertex(v);
        validateVertex(w);
        return id[v] == id[w];
    }

    /**
     * 检查某个顶点的连通分量的标识符
     *
     * @param v 指定的顶点
     * @return 标识符
     */
    public int id(int v) {
        validateVertex(v);
        return id[v];
    }

    /**
     * 图的访问次数
     *
     * @return 访问次数
     */
    public int count() {
        return count;
    }

    /**
     * 获取指定顶点的联通顶点数量
     *
     * @param v  指定的顶点
     * @return 顶点的数量
     */
    public int size(int v) {
        validateVertex(v);
        return size[id[v]];
    }

    /**
     * 验证顶点
     *
     * @param v 验证的顶点
     */
    private void validateVertex(int v) {
        int V = marked.length;
        if (v < 0 || v >= V) {
            throw new IllegalArgumentException("vertex " + v + " is not between 0 and " + (V - 1));
        }
    }

    /**
     * 单元测试
     *
     * @param args 命令行参数
     */
    public static void main(String[] args) {
        In in = new In(args[0]);
        Graph G = new Graph(in);
        ConnectedComponent cc = new ConnectedComponent(G);

        int m = cc.count();
        StdOut.println(m + " components");

        Queue<Integer>[] components = (Queue<Integer>[]) new Queue[m];
        for (int i = 0; i < m; i++) {
            components[i] = new Queue<>();
        }
        for (int v = 0; v < G.V(); v++) {
            components[cc.id(v)].enqueue(v);
        }

        for (int i = 0; i < m; i++) {
            for (int v : components[i]) {
                StdOut.print(v + " ");
            }
            StdOut.println();
        }
    }
}

File: Kotlin/src/me/metatom/kotlin/graph/ConnectedComponent.kt
package me.metatom.kotlin.graph

import edu.princeton.cs.algs4.EdgeWeightedGraph
import edu.princeton.cs.algs4.Graph
import edu.princeton.cs.algs4.In
import edu.princeton.cs.algs4.Queue

/**
 * 使用深度优先搜索查找图中的连通分量
 *
 * @author igaozp
 * @since 2017-09-08
 * @version 1.0
 */
class ConnectedComponent {
    // 存储图中顶点的访问状态
    private var marked = MutableList(0) { false }
    // 连通分量的标识符
    private var id = MutableList(0) { 0 }
    // 存放给定分量的顶点数量
    private var size = MutableList(0) { 0 }
    // 顶点的访问次数
    private var count: Int = 0

    constructor(G: Graph) {
        marked = MutableList(G.V()) { false }
        id = MutableList(G.V()) { 0 }
        size = MutableList(G.V()) { 0 }

        for (s in 0 until G.V()) {
            if (!marked[s]) {
                dfs(G, s)
                count++
            }
        }
    }

    constructor(G: EdgeWeightedGraph) {
        marked = MutableList(G.V()) { false }
        id = MutableList(G.V()) { 0 }
        size = MutableList(G.V()) { 0 }

        for (v in 0 until G.V()) {
            if (!marked[v]) {
                dfs(G, v)
                count++
            }
        }
    }

    // 深度优先搜索
    private fun dfs(G: Graph, v: Int) {
        marked[v] = true
        id[v] = count
        size[count]++

        // 访问与 v 连通的顶点
        G.adj(v)
                .filterNot { marked[it] }
                .forEach { dfs(G, it) }
    }

    // 深度优先搜索
    private fun dfs(G: EdgeWeightedGraph, v: Int) {
        marked[v] = true
        id[v] = count
        size[count]++

        // 访问与 v 连通的顶点
        G.adj(v)
                .filterNot { marked[it.other(v)] }
                .forEach { dfs(G, it.other(v)) }
    }

    // 检查两个连通分量是否连通
    fun connected(v: Int, w: Int): Boolean {
        validateVertex(v)
        validateVertex(w)
        return id[v] == id[w]
    }

    // 检查某个顶点的连通分量的标识符
    fun id(v: Int): Int {
        validateVertex(v)
        return id[v]
    }

    // 图的访问次数
    fun count(): Int = count

    // 获取指定顶点的联通顶点数量
    fun size(v: Int): Int {
        validateVertex(v)
        return size[id[v]]
    }

    // 验证顶点
    fun validateVertex(v: Int) {
        val V = marked.size
        if (v < 0 || v >= V) {
            throw IllegalArgumentException("vertex $v is not between 0 and ${V - 1}")
        }
    }
}

// 单元测试
fun main(args: Array<String>) {
    val input = In(args[0])
    val G = Graph(input)
    val cc = ConnectedComponent(G)

    println("${cc.count()} components")

    val components = MutableList(cc.count(), { Queue<Int>() })

    for (v in 0 until G.V()) {
        components[cc.id(v)].enqueue(v)
    }

    for (i in 0 until cc.count()) {
        components[i].forEach { println("$it ") }
        println()
    }
}